### PR TITLE
Bugfix in getMaxReadType for type variables.

### PR DESCRIPTION
**Pull request by bgogul**
This PR ensures that, in the returned the max read type for a type variable, the `writeSuperset` is consistent with the `readSubset`.

### CODE REVIEW

File: src/runtime/policy/ingress-validation.ts [hunk 1]
```diff
@@ -294,18 +294,29 @@ export class IngressValidation {
...
-            /* canWriteSuperset = */ maxReadType,
-            /* canReadSubset = */ canReadSubset,
-            typeVar.resolveToMaxType
-          );
+          newCanWriteSuperset = maxReadType.restrictTypeRanges(canReadSubset);
         }
+        return TypeVariable.make(
+          '',
+          /* canWriteSuperset = */ newCanWriteSuperset,
+          /* canReadSubset = */ canReadSubset,
```
Comment (mmandlis) [inline src/runtime/policy/ingress-validation.ts:317]: nit = comment seems redundant here

File: src/runtime/policy/tests/policy-test.ts [hunk 2]
```diff
@@ -808,9 +808,97 @@ policy MyPolicy {
...
+    `)).schemas['A']
+    const writeEntityType = new EntityType(writeA);
+    const typeVar = TypeVariable.make(
+         '',
+         /* canWriteSuperset = */null,
+         /* canReadSubset = */writeEntityType);
+    // The expected max read type variable should have `A { foo {a} }` for writeSuperset.
+    const maxReadA = (await Manifest.parse(`
+      schema A
+        // Should be the following when b/175169555 is fixed:
```
Comment (mmandlis) [inline src/runtime/policy/tests/policy-test.ts:887]: maybe rephrase to `// TODO(b/...): ...` so it's easier to find when the bug is fixed?

File: src/runtime/policy/tests/policy-test.ts [hunk 3]
```diff
@@ -808,9 +808,97 @@ policy MyPolicy {
...
-       assert.deepEqual(ingressValidation.getMaxReadType(subsetTypeVar), subsetTypeVar);
+       const expectedTypeVar = TypeVariable.make(
+         '',
+         /* canWriteSuperset = */personSubset,
+         /* canReadSubset = */personSubset);
+       assert.deepEqual(
+         ingressValidation.getMaxReadType(subsetTypeVar), expectedTypeVar);
      });
 
+  it('updates write w/ max read type that is consistent with read', async() => {
```
Comment (mmandlis) [inline src/runtime/policy/tests/policy-test.ts:819]: there are 3 possible cases in the updated method above:
```
if (maxReadType.isAtLeastAsSpecificAs(canReadSubset)) {
    newCanWriteSuperset = canReadSubset;
} else if (canReadSubset.isAtLeastAsSpecificAs(maxReadType)) {
    newCanWriteSuperset = maxReadType;
} else {
    return TypeVariable.make(...)
}
```
are there unit tests for each?

File: src/runtime/policy/ingress-validation.ts [hunk 4]
```diff
@@ -294,18 +294,29 @@ export class IngressValidation {
...
+        // TODO(b/175169555): Schema.intersect has a bug with respect to inline
+        // entites. The following code works around the bug and when it is fixed,
+        // it should be simplified as follows:
+        //   const canWriteSuperset = maxReadType.restrictTypeRanges(canReadSubset);
+        var newCanWriteSuperset = null
         if (maxReadType.isAtLeastAsSpecificAs(canReadSubset)) {
-          // Write constraint on the type variable indicates that already only
-          // the allowed fields are written; return the type variable itself.
-          return type;
+          newCanWriteSuperset = canReadSubset
```
Comment (mmandlis) [inline src/runtime/policy/ingress-validation.ts:308]: nit - ; missing at the end of line here and below.

File: src/runtime/policy/tests/policy-test.ts (hunk 3 above)
Comment (bgogul) [reply src/runtime/policy/tests/policy-test.ts]: Yes. The following tests cover the other cases:
 - https://github.com/PolymerLabs/arcs/blob/0af125651abc4e3bc1684522a366b27c5459f16f/src/runtime/policy/tests/policy-test.ts#L746
  - https://github.com/PolymerLabs/arcs/blob/0af125651abc4e3bc1684522a366b27c5459f16f/src/runtime/policy/tests/policy-test.ts#L778

File: src/runtime/policy/tests/policy-test.ts (hunk 2 above)
Comment (bgogul) [reply src/runtime/policy/tests/policy-test.ts]: Done.

File: src/runtime/policy/ingress-validation.ts (hunk 1 above)
Comment (bgogul) [reply src/runtime/policy/ingress-validation.ts]: done.

File: src/runtime/policy/ingress-validation.ts (hunk 4 above)
Comment (bgogul) [reply src/runtime/policy/ingress-validation.ts]: fixed a few more lint errors.

**Comment by bgogul** [comment]
Thanks for the review, @mmandlis I addressed your comments and questions. PTAL.

File: src/runtime/policy/tests/policy-test.ts [hunk 5]
```diff
@@ -808,9 +808,97 @@ policy MyPolicy {
...
+    // Note that the max read type for schema A is `A {a, b}`. However, the
+    // expected max read type variable should have `A {a}` for writeSuperset
+    // so that it is consistent with the write `A {a, d}`.
+    const maxReadA = (await Manifest.parse(`
+      schema A
+        a: Text
+    `)).schemas['A'];
+    deleteFieldRecursively(maxReadA, 'location', {replaceWithNulls: true});
+    const maxReadEntityType = new EntityType(maxReadA);
+    const expected = TypeVariable.make(
```
Comment (mmandlis) [inline src/runtime/policy/tests/policy-test.ts:853]: optional- 
the test is a bit hard to follow. maybe create a helper method, smth like:

`createTypeVar(canWriteSupersetSchema: string, canReadSebsetSchema: string): TypeVariable`

that can be called like:
`createTypeVar( 'a: Text', 'a: Text, d: Text')`

**Review by mmandlis — APPROVED** [review]

**Review by piotrswigon — APPROVED** [review]
Looks legit, but honestly I haven't spent enough time to fully parse and internalize the logic.

File: src/runtime/policy/tests/policy-test.ts (hunk 5 above)
Comment (bgogul) [reply src/runtime/policy/tests/policy-test.ts]: Excellent Idea. Done.
